common/src/cost_model.rs

```diff
@@ -0,0 +1,465 @@
+
+    fn try_from(db_model: DbCostModel) -> Result<Self, Self::Error> {
+        Ok(Self {
+            deployment: DeploymentId::from_str(&db_model.deployment)?,
+            model: db_model.model,
+            variables: db_model.variables,
+        })
+    }
+}
+
+impl From<CostModel> for DbCostModel {
+    fn from(model: CostModel) -> Self {
+        let deployment = model.deployment;
+        DbCostModel {
+            deployment: format!("{deployment:#x}"),
+            model: model.model,
+            variables: model.variables,
+        }
+    }
+}
+
+/// Query cost models from the database, merging the global cost model in
+/// whenever there is no cost model defined for a deployment.
+pub async fn cost_models(
+    pool: &PgPool,
+    deployments: &[DeploymentId],
+) -> Result<Vec<CostModel>, anyhow::Error> {
+    let hex_ids = deployments
+        .iter()
+        .map(|d| format!("{d:#x}"))
+        .collect::<Vec<_>>();
+
+    let mut models = if deployments.is_empty() {
+        sqlx::query_as!(
+            DbCostModel,
+            r#"
+            SELECT deployment, model, variables
+            FROM "CostModels"
+            WHERE deployment != 'global'
+            ORDER BY deployment ASC
+            "#
+        )
+        .fetch_all(pool)
+        .await?
+    } else {
+        sqlx::query_as!(
+            DbCostModel,
+            r#"
+            SELECT deployment, model, variables
+            FROM "CostModels"
+            WHERE deployment = ANY($1)
+            AND deployment != 'global'
+            ORDER BY deployment ASC
+            "#,
+            &hex_ids
+        )
+        .fetch_all(pool)
+        .await?
+    }
+    .into_iter()
+    .map(CostModel::try_from)
+    .collect::<Result<Vec<_>, _>>()?;
+
+    let deployments_with_models = models
+        .iter()
+        .map(|model| &model.deployment)
+        .collect::<HashSet<_>>();
+
+    let deployments_without_models = deployments
+        .iter()
+        .filter(|deployment| !deployments_with_models.contains(deployment))
+        .collect::<HashSet<_>>();
+
+    // Query the global cost model
+    if let Some(global_model) = global_cost_model(pool).await? {
+        // For all deployments that have a cost model, merge it with the global one
+        models = models
+            .into_iter()
+            .map(|model| merge_global(model, &global_model))
+            // Inject a cost model for all deployments that don't have one
+            .chain(
+                deployments_without_models
+                    .into_iter()
+                    .map(|deployment| CostModel {
+                        deployment: deployment.to_owned(),
+                        model: global_model.model.clone(),
+                        variables: global_model.variables.clone(),
+                    }),
+            )
+            .collect();
+    }
+
+    Ok(models)
+}
+
+/// Make database query for a cost model indexed by deployment id
+pub async fn cost_model(
```

Comment:
Same thing. I think this function does too many different things.

Reply:
Same here.